<h2 align="center">Brute force vs. exponentiating by squaring</h2>

<br>

Let's examine the performance of two different algorithms for solving the following problem:

$$ a^n = a \, \cdot \, a \, \cdot \, ... \, \cdot \, a. $$

In [1]:
import time

<h3 align="left">Brute force algorithm</h3>

Brute force method refers to the straightforward solution of a problem. The algorithms produced by this method are typically simple and straightforward, but not necessarily very efficient.

In [10]:
def brute_force(a, n):
    """
    Calculate the product a^n using the brute force method.
    
    Args:
        a (int): The base of an exponent.
        n (int): The number of times the base is multiplied by itself.
        
    Returns:
        int: The product of a^n, where a is the base and n is the exponent.
    """
    # Everything to the power of zero equals 1 (except 0^0, which is undefined).
    if n == 0:
        return 1
    
    if n < 0:
        a = 1/a
        n = -n
    
    # A base of zero always returns zero (0^0 excluded)
    if a == 0:
        # Undefined
        if n == 0:
            raise ValueError("Undefined result, please provide valid base and exponent values.")
        else:
            return 0
    
    # Create a helper variable that is used to store the result
    result = 1
    # Use a for loop to calculate the product a^n
    for _ in range(n):
        result *= a
    
    return result

<h3 align="left">Exponentiating by Squaring algorithm</h3>

Exponentiation by squaring is an algorithm that allows one to compute the power $\, a^n \,$ more efficiently than the brute force method. The basic idea behind exponentiation by squaring is to take advantage of the fact that $\, a^n \,$ can be expressed differently based on whether n is even or odd:

$$ a^n = \left\{
\begin{array}{ll}
        1, & \text{if n = 0} \\
        \frac{1}{a^n}, & \text{if n < 0} \\
        (a^{n/2})^2, & \text{if n is even} \\
        a \cdot \left(a^{\frac{n-1}{2}} \right)^2, & \text{if n is odd}
\end{array}
\right. $$

This recursive approach reduces the number of multiplications needed to compute the result.

In [3]:
def exp_by_squaring(a: int, n: int):
    """
    Calculate the product a^n using the exponentiation by squaring method.
    
    Args:
        a (int): The base of an exponent.
        n (int): The number of times the base is multiplied by itself.
        
    Returns:
        int: The product of a^n, where a is the base and n is the exponent.
    """
    if n == 0:
        return 1
    
    if n < 0:
        a = 1/a
        n = -n
    
    if n == 1:
        return a
    
    if n == 2:
        return a*a
    
    # Exponent is even
    if n % 2 == 0:
        exp_halved = exp_by_squaring(a, n/2)
        result = exp_halved**2
    
    # Exponent is odd
    else:
        exp_halved = exp_by_squaring(a, (n-1)/2)
        result = a * exp_halved**2
    
    return result

Now, let us compare the performance of the **brute force** and **exponentiation by squaring** *algorithms*.

As one would expect, the functions return the same result:

In [4]:
brute_force(10, 1000) == exp_by_squaring(10, 1000)

True

However, the performance of the algorithms is not the same.

In [5]:
def brute_force_performance(a: int, n: int):
    """
    Measure the performance of the brute_force() function for calculating a^n.

    Args:
        a (int): The base of an exponent.
        n (int): The number of times the base is multiplied by itself.

    Returns:
        str: A string indicating the execution time of the brute_force() function.
    """
    # Record the start time
    start_time = time.time()
    
    # Call the brute_force() function
    brute_force(a, n)
    
    # Record the end time
    end_time = time.time()
    
    return f'Execution time: {end_time - start_time} seconds.'

In [6]:
def exp_by_squaring_performance(a:int, n: int):
    """
    Measure the performance of the exp_by_squaring() function for calculating a^n.

    Args:
        a (int): The base of an exponent.
        n (int): The number of times the base is multiplied by itself.

    Returns:
        str: A string indicating the execution time of the exp_by_squaring() function.
    """
    # Record the start time
    start_time = time.time()
    
    # Call the exp_by_squaring() function
    exp_by_squaring(a, n)
    
    # Record the end time
    end_time = time.time()
    
    return f'Execution time: {end_time - start_time} seconds.'

In [127]:
brute_force_performance(10, 1000)

'Execution time: 0.0002231597900390625 seconds.'

In [125]:
exp_by_squaring_performance(10, 1000)

'Execution time: 3.0040740966796875e-05 seconds.'

- The brute force algorithm calculates $\, 10^{1000} \,$ in 0.0002231597900390625 seconds


- The exponentiation by squaring algorithm calculates $\, 10^{1000} \,$ in 0.000030040740966796875 seconds.

In [133]:
0.0002231597900390625 / 3.0040740966796875e-05

7.428571428571429

- The time it took for the **brute force** algorithm to calculate $\, 10^{1000} \,$ was approximately **7.429** times greater than the time it took for the **exponentiation by squaring** *algorithm* to calculate the same product.

- In other words, the **exponentiation by squaring** algorithm performed approximately **7.429** times faster than the **brute force** algorithm, when a=10 and n=1000.